### ISSUE TITLE
Failed when testing 138_7.JPG in test_toy file 

### ISSUE THREAD

**Issue by Chokurei** [bsun0802/Zero-DCE#1]
Thank you so much for the provided code! Here I found some issues when configuring the code and data.
1. in test_toy file, image 138_7.JPG can only get over-exposure result while using your provided trained model 512-10Wexp-1_ckpt.pth, here is the result:
original: ![138_7_original](https://user-images.githubusercontent.com/16301109/81491835-b3294a00-92cd-11ea-92ac-7f3a949573d3.jpg)
enhanced: ![138_7_enhanved](https://user-images.githubusercontent.com/16301109/81491841-c0decf80-92cd-11ea-9b78-2615359feb3c.jpg)  
I wonder the trained model is only suitable for under-exposure imagery. 

2. I load your model by eval.py; it reveals that the parameter 'n_LE' you chose is 2, am I right? as follows:
![res_111_1](https://user-images.githubusercontent.com/16301109/81491898-39de2700-92ce-11ea-9e4b-e44498a0ddad.jpg)

3. When I trained my model, it looks that they cannot work at all. So I wonder what epoch number you chose, and would you kindly tell me what may cause such failure? here is my failed result example:
![res_106_1](https://user-images.githubusercontent.com/16301109/81493105-875f9180-92d8-11ea-9261-7d87f5507663.jpg)

The hp I used is
```python
hp = dict(lr=1e-4, wd=0.0, lr_decay_factor=0.9,
          n_LE=8, std=0.05,
          w_exp=1, w_col=0.5, w_tvA=20,
          spa_rsize=4, exp_rsize=16)
```
 
The model is trained by 
python train.py --device 1 --experiment trial --baseDir ../data/part1-512 --numEpoch 300
  
Look forward to your kind answer.

**Comment by bsun0802** [comment]
Thanks for starring my repo. To answer your questions, 

1. 
Indeed n_LE=2 is a toy example, the n_LE=2 model I uploaded only learns to enhance under-exposed images. Indeed if you examine the A values, all of them will be near 1, so the model is indeed applying y = 2x - x^2 = for all every pixel x (and y>x). 

2.
I updated this repository completely, and also uploaded another pytorch model file which I trained with n_LE=8. You can try it with the new `demo.py`. 

3.
The hyper-parameters as well as the number of epochs are saved in the pytorch checkpoint as well, you can access them as `checkpoint['HyperParam']` and `checkpoint['epoch']` after you load the model into a variable named `checkpoint`.

However, when n_LE=8, I found it hard to train and balance the 4 losses described in the paper. w_col=0.5 and w_tvA=20 may not apply here, because we don't know how the author implemented those losses. I add arguments for the weight of each loss in the new `train.py`, and you can find the weights I used in the new .pth model I uploaded.

**Comment by Chokurei** [comment]
Thank you so much for your quick reply. I got it!
When I checked your final results, actually there are many artifacts around over/under exposure regions; and I also find that it's too hard to train the promising model as the author mentioned in the original paper. If you have any progress please kindly share it with me.

**Comment by bsun0802** [comment]
Hi, Zhiling,

In order to deal with the color artifact, I devised my own color constancy loss to replace the one described in the paper. 

And I obtain noticeable improvement by doing so.  I updated the repo to reflect all the changes and uploaded the model checkpoint. 

One more barrier remained is that, `edges` especially `edges with pure color` will have a halo around them, I think the spatial loss and total variation loss are responsible for that. But I can not figure out how to deal with it yet, but the total variation loss is indeed share some resemblance in the operator used in edge detection.  Maybe you can tune the model for more to figure it out and good luck!

**Comment by Chokurei** [comment]
Thanks for sharing the update! I have tried the new model '4LE-0_best_model.pth' trained by your color constancy loss, the results get better indeed. Compared with other paired_data- referenced papers, Zero-reference seems relatively elegant. I will try to figure out the reason for the left artifacts based on your suggestion.

**Comment by bsun0802** [comment]
Hi,  

I've made progress in ZeroDCE training, don't know if you are still interested in it. 

In terms of eliminating the artefact that we have been seeing. I found setting the E to be 0.7 helps a lot (I was defaultly believing the author saying E in [0.4, 0.7] does not affect result and set it to be 0.6, but it was not the case).
